tests/test_directives.py

```diff
@@ -0,0 +1,270 @@
+        product_directive.serial(10, 20) == 10
+
+    def test_manual_item_parallel(self, product_directive):
+        product_directive.parallel(10, 20) == 10
+
+    def test_manual_item_finalize(self, product_directive):
+        asset_dict = {'free': False, 'discount': 5}
+        assert product_directive.finalize(50, asset_dict) == 45
+        asset_dict['free'] = True
+        assert product_directive.finalize(50, asset_dict) == 0
+
+
+class TestDirectives:
+    """
+    Tests for Directives Class
+    """
+    def test_get_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        for item in available_directives_list:
+            assert directives[item.name] == item.default
+
+    def test_add_directive(self, available_directives_list):
+        directives = Directives(available_directives_list[:-1])
+        directives._add_directive(PROCESS_FRACTION)
+        assert directives[PROCESS_FRACTION.name] == PROCESS_FRACTION.default
+        with pytest.raises(TypeError):
+            directives._add_directive('Test')
+        with pytest.raises(ValueError):
+            directives._add_directive(PROCESS_FRACTION)
+
+    def test_set_defined_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives._set_defined_directive(NP.name, 10)
+        assert directives[NP.name] == 10
+
+    def test_set_defined_directive_invalid(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        with pytest.raises(ValueError):
+            directives._set_defined_directive(NP.name, 0)
+
+    def test_set_undefined_directive(self, available_directives_list, product_directive):
+        directives = Directives(available_directives_list)
+        with pytest.raises(Exception):
+            directives._set_defined_directive(product_directive.name, 0)
+
+    def test_set_directives_item(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives['test'] = True
+        assert directives['test']
+
+    def test_del_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives['test'] = True
+        directives._set_defined_directive(NP.name, 100)
+        assert directives[NP.name] == 100
+        assert directives['test']
+        del directives[NP.name]
+        assert directives[NP.name] == NP.default
+        del directives['test']
+        with pytest.raises(KeyError):
+            directives['test']
+
+    def test_update_directive_without_aggregate(
+        self, available_directives_list, non_default_directive_values
+    ):
+        directives1 = Directives(available_directives_list)
+        valid_values_0 = non_default_directive_values()
+        valid_values_1 = non_default_directive_values(1)
+        for dirs in directives1:
```

Comment:
I would use `directives1.update(valid_values_0)` here too, and just test the assert. Also with this update, you shouldn't have to created an `expected_values` `dict` right? The expected values are just the values in the `valid_values_{0,1}`.

Reply:
Actually the `__getitem__` function will use the environment directives' `finalize` methods, so the returned values will not be identical. However, the is no real reason to set `directives1` to the values in `non_default_directives_values[0]` and then update it to one.

Reply:
Oh okay, now I get your point